WcaOnRails/app/models/person.rb

```diff
@@ -167,11 +167,66 @@ def best_solve(event, type)
   end
 
   def world_championship_podiums
-    results.includes(:competition, :event, :format)
-           .joins(competition: [:championships])
-           .podium
+    results.podium
+           .joins(:event, competition: [:championships])
            .where("championships.championship_type = 'world'")
            .order("year DESC, Events.rank")
+           .includes(:competition, :format)
+  end
+
+  def championship_podiums_with_condition
+    # Get all championship competitions of the given type where the person made it to the finals.
+    # For each of these competitions, get final results only for people eligible for the championship
+    # and reassign their positions. If a result belongs to the person, add it to the array.
+    [].tap do |championship_podium_results|
+      yield(results)
+        .final
+        .succeeded
+        .order("year DESC")
+        .includes(:competition)
+        .map(&:competition)
+        .uniq
+        .each do |competition|
+          yield(competition.results)
+            .final
+            .succeeded
+            .joins(:event)
+            .order("Events.rank, pos")
+            .includes(:format, :competition)
+            .group_by(&:eventId)
+            .each_value do |final_results|
+              positions_delta = 0
+              previous_pos = 0
+              final_results.each_with_index do |result, index|
+                positions_delta += (previous_pos == result.pos ? 0 : previous_pos - result.pos + 1)
+                previous_pos = result.pos
+                result.pos += positions_delta
+                break if result.pos > 3
+                championship_podium_results.push result if result.personId == self.wca_id
+              end
+            end
+        end
+    end
+  end
+
+  def championship_podiums
+    {}.tap do |podiums|
+      podiums[:world] = world_championship_podiums
+      podiums[:continental] = championship_podiums_with_condition do |results|
+        results.joins(:country, competition: [:championships]).where("championships.championship_type = Countries.continentId")
+      end
+      EligibleCountryIso2ForChampionship.where(eligible_country_iso2: self.country_iso2).distinct.pluck(:championship_type).each do |championship_type|
```

Comment:
Oh wow, what an edge case... So you're right, if we had two different custom championship types, and then someone switches his country so that he is eligible to win the other one, he would be treated as eligible to win both.

Reply:
Fortunately fixing this behaviour is extremely easy, if we do this for *all* championship types, then it should work just fine!

Reply:
I was actually worried about the reverse: someone who used to represent China now represents the US, and `self.country_iso2` is equal to US for them, so they aren't considered eligible for past Greater China Championships they attended. The new code looks good to me!

Reply:
Oh wow, so tricky that my first comment wasn't even true.